esp32: add support for esp-idf bootloader, XIP execution and linker script updates

sylvioalves opened the pull request:
This commit is part of the initial [ESP32 Development](https://github.com/zephyrproject-rtos/zephyr/issues/29394) work.

mahavirj committed:
boards: esp32: add XIP support and enable bootloader build

Disable RTC WDT enabled (by default) by 2nd stage bootloader in ESP-IDF.
This WDT timer ensures correct hand-over and startup sequence from
bootloader to application.

Enabling bootloader caused system clock initialization to fail
when clock rate is greater then 80MHz. This also fixes
esp32 clock source code.

Signed-off-by: Mahavir Jain <mahavir@espressif.com>

shubhamkulkarni97 committed:
boards: esp32: add 'nostdinc' expection for espressif toolchain

disable nostdinc for espressif tolchain

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

shubhamkulkarni97 committed:
boards: esp32: linker: move libraries and reserve DRAM regions

Fix issues with restoring symbols from common ld templates
Workaround esptool linker sections limit
Move kernel library into IRAM
Improve UDP throughput

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

sylvioalves committed:
drivers: entropy: esp32: update register call

Update entropy driver to use proper registers

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

shubhamkulkarni97 committed:
boards: esp32: use cmake to build ESP-IDF bootloader

Make builds are not supported on Windows command prompt

Signed-off-by: Shubham Kulkarni <shubham.kulkarni@espressif.com>

sylvioalves committed:
docs: esp32: updated environment setup

Previous Zephyr environment required the usage
of external ESP-IDF cloning as source of headers
and libraries. The latest implementation uses external
module as source.

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

sylvioalves committed:
west.yml: update esp32 manifest

Update esp32 hal driver
Added west-commands option to allow proper
submodules download

Signed-off-by: Sylvio Alves <sylvio.alves@espressif.com>

carlescufi commented:
@sylvioalves could you please rebase this on top of the latest master and force push?

sylvioalves commented:
> @sylvioalves could you please rebase this on top of the latest master and force push?

Sure. However, there too many check errors that I need to handle, most with code style.

carlescufi commented:
> > @sylvioalves could you please rebase this on top of the latest master and force push?
> 
> Sure. However, there too many check errors that I need to handle, most with code style.

Thanks! I fixed the GitHub action that posts the manifest update comment now.

sylvioalves commented:
@carlescufi Build check won't pass until this [PR](https://github.com/zephyrproject-rtos/esp-idf/pull/6/) gets merged.

dcpleung commented on arch/xtensa/core/CMakeLists.txt:
```diff
@@ -16,6 +16,9 @@ zephyr_library_sources(
 zephyr_library_sources_ifndef(CONFIG_ATOMIC_OPERATIONS_C atomic.S)
 zephyr_library_sources_ifdef(CONFIG_XTENSA_USE_CORE_CRT1 crt1.S)
 zephyr_library_sources_ifdef(CONFIG_IRQ_OFFLOAD irq_offload.c)
-zephyr_library_sources_ifdef(CONFIG_THREAD_LOCAL_STORAGE tls.c)
```
Please do not remove this.

dcpleung commented on arch/xtensa/core/fatal.c:
```diff
@@ -92,8 +98,12 @@ void z_xtensa_fatal_error(unsigned int reason, const z_arch_esf_t *esf)
 		z_xtensa_dump_stack(esf);
 	}
-
-	z_fatal_error(reason, esf);
```
`z_fatal_error()` does not always halt the system. Applications can implement their own `k_sys_fatal_error_handler()` to override the default behavior. So you need to keep this and remove the `arch_system_halt()` in new code.

sylvioalves replied:
I dropped the backtrace commit from this PR and will fix in another PR.

dcpleung commented on arch/xtensa/include/xtensa-asm2-context.h:
```diff
@@ -78,17 +79,11 @@
 #define BASE_SAVE_AREA_SIZE_SCOMPARE	0
 #endif
 
-#if XCHAL_HAVE_THREADPTR && defined(CONFIG_THREAD_LOCAL_STORAGE)
```
Why are you removing all the TLS related bits?

carlescufi commented:
> PR

I've commented on that PR, there's something wrong with the commit set (8K) in there, you probably need to rebase it.

sylvioalves commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -18,29 +19,55 @@
 PROVIDE ( esp32_rom_ets_set_appcpu_boot_addr = 0x4000689c );
 PROVIDE ( esp32_rom_i2c_readReg = 0x40004148 );
 PROVIDE ( esp32_rom_i2c_writeReg = 0x400041a4 );
+PROVIDE ( phy_get_romfuncs = 0x40004100 );
```
Yes, we plan to update this to `esp_rom_`  soon. For now, we leave this way.

sylvioalves commented on dts/xtensa/espressif/esp32.dtsi:
```diff
@@ -126,7 +126,7 @@
 			/* interrupts = <33 0>; - FIXME: Enable interrupts when interrupt-controller got supported in device tree */
 			label = "TRNG_0";
-			status = "disabled";
+			status = "okay";
```
It could be disabled by default. We let this enable for wifi library that is to be pushed.

dcpleung commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -19,28 +19,31 @@
 PROVIDE ( esp32_rom_ets_set_appcpu_boot_addr = 0x4000689c );
 PROVIDE ( esp32_rom_i2c_readReg = 0x40004148 );
 PROVIDE ( esp32_rom_i2c_writeReg = 0x400041a4 );
+PROVIDE ( ets_printf = 0x40007d54 );
```
Since this is provide by the ROM, having a prefix of `esp32_rom_` would make it consistent with other functions here.

sylvioalves replied:
@dcpleung, yes, it makes sense. However, to make it easier for future esp-idf rebase and updates, we woud like to maintain consistence. Soon we will update this.

pabigot replied:
Please fix it before this PR is merged.  The name exposed by the Zephyr linker script should not need to be the same as the name used in the ROM, nor the name used in `esp-idf`.

One way to make this cleaner is to add the same function with two aliases:
* one using the spelling used in `esp-idef` until it's cleaned up;
* another with the `esp32_rom_` prefix used in Zephyr

Use in Zephyr should use the name with the proper prefix, and the secondary alias should have a comment explaining why it's there and the conditions to be met so it can be removed.  All those secondary aliases could be grouped together to help make that clear.

dcpleung commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -19,28 +19,31 @@
 PROVIDE ( esp32_rom_i2c_readReg = 0x40004148 );
 PROVIDE ( esp32_rom_i2c_writeReg = 0x400041a4 );
+PROVIDE ( ets_printf = 0x40007d54 );
+PROVIDE ( intr_matrix_set = 0x4000681c );
```
Just wondering why are you renaming this? This is provided by the ROM and have the same address as before.

sylvioalves replied:
@dcpleung same reason as previous comment.

nashif commented on west.yml:
```diff
@@ -43,8 +43,9 @@ manifest:
+    - name: hal_espressif
+      west-commands: west/west-commands.yml
+      revision: c836ae28675a7ce8b4bdb75b7379fdccfe2731fb
       path: modules/hal/esp-idf
```
should point to a new directory, not esp-idf, maybe just espressif like other HALs

sylvioalves replied:
yes, thanks.

sylvioalves commented:
Hi, [hal_espressif](https://github.com/zephyrproject-rtos/hal_espressif) repo is in place and supporting this PR.  
Any extra reviews are welcome.

nashif commented on boards/xtensa/esp32/doc/index.rst:
```diff
@@ -48,21 +48,56 @@ System requirements
+.. code-block:: console
+
+   # On Linux and macOS:
+   export ESP_IDF_PATH=<zephyr workspace top>/modules/hal/esp-idf"
```
do you still need that? we know this path via ZEPHYR_ESPRESSIF_MODULE_DIR already so I do not think this should be set

sylvioalves replied:
of course, this is not needed anymore.

nashif commented on boards/xtensa/esp32/doc/index.rst:
```diff
@@ -48,21 +48,56 @@ System requirements
+   export ESP_IDF_PATH=<zephyr workspace top>/modules/hal/esp-idf"
+
+   # On Windows:
+   set ESP_IDF_PATH=<zephyr workspace top>\modules\hal\esp-idf"
```
this should be hal_espressif now

sylvioalves replied:
In fact, it is "espressif" now. `modules/hal/espressif`.

I will have to adress the path in some way due to the need of `./install.sh` and `source export.sh` within modules folder.

sylvioalves commented:
@nashif, for the latest update:

- python env is not necessary anymore. It means `west` does not need reinstalling.
- fixed windows cmake issues regarding the python clean up
- updated getting started document describing OSes configuration

For esp32 bring up, one's need only to run 

`west update`
`west espressif update`

and export necessary environment variables as described in the getting started tutorial.

tejlmand commented on boards/xtensa/esp32/CMakeLists.txt:
```diff
@@ -4,28 +4,44 @@
-    make
-    IDF_PATH=${ESP_IDF_PATH}
-    BUILD_DIR_BASE=${espidf_build_dir}
-    SDKCONFIG=${espidf_build_dir}/sdkconfig
-    defconfig bootloader partition_table
+    ${espidf_tool}
+    cmake -GNinja
+    -S ${espidf_components_dir}/bootloader/subproject
+    -B ${espidf_build_dir}/bootloader -DSDKCONFIG=${espidf_build_dir}/sdkconfig
+    -DIDF_PATH=${ESP_IDF_PATH} -DIDF_TARGET=${CONFIG_BOARD}
+    -DPYTHON_DEPS_CHECKED=1
+    && cd ${espidf_build_dir}/bootloader && cmake --build .
```
This should be divided into:
- `CONFIGURE_COMMAND`
- `BUILD_COMMAND`

also use `${CMAKE_COMMAND}` instead of `cmake` and `${CMAKE_GENERATOR}` instead of Ninja, to so that same generator is used for both current project and the external one.

There is no need for the `cd` command, as the build command will execute fot the `BINARY_DIR` var.

```suggestion
    CONFIGURE_COMMAND
    ${espidf_tool}
    ${CMAKE_COMMAND} -G${CMAKE_GENERATOR}
    -S ${espidf_components_dir}/bootloader/subproject
    -B ${espidf_build_dir}/bootloader -DSDKCONFIG=${espidf_build_dir}/sdkconfig
    -DIDF_PATH=${ESP_IDF_PATH} -DIDF_TARGET=${CONFIG_BOARD}
    -DPYTHON_DEPS_CHECKED=1
    BUILD_COMMAND
    ${CMAKE_COMMAND} --build .
```

pabigot commented on boards/xtensa/esp32/doc/index.rst:
```diff
@@ -48,30 +48,48 @@ System requirements
-function.
+In order to build this port, the `SDK`_ and the `toolchain`_ need
+to be manually addressed. Both are provided by the SoC manufacturer.
+Cloning ESP_IDF is not necessary as it is automatically fetched in external modules.
```
From what I can tell, this would install in the location recommended in the upstream documentation, because `west expressif update` for me did a few things then said the toolchains were already installed.  If so, please document that: some of us would get upset when things are silently installed in `${HOME}`.

Because that installation done behind the scenes with the west command it might be helpful to change the text below that's currently:

    export ESPRESSIF_TOOLCHAIN_PATH="/path/to/xtensa-esp32-elf/"

to give a better hint of what it needs to be provided.  In my case it's probably:

    ${HOME}/.espressif/tools/xtensa-esp32-elf/esp-2020r3-8.4.0/xtensa-esp32-elf/

which is a mouthful.  It also looks like that would have to change for s2 and s3 as those seem to have different toolchains accessed through a different mid-level path component.

sylvioalves replied:
@pabigot, If I understand:
1) Write down on the getting started tutorial what `west espressif update` will perform in details;
2) Change the export statement regarding xtensa toolchain;

As we have no support yet for S2 and S3, I think we can keep `(...)/xtensa-esp32-elf/` as is. Ok?

pabigot replied:
Sure.  But if it's true that supporting esp32s2 and esp32s3 would require changing the `EXPRESSIF_TOOLCHAIN_PATH` depending on the target it would be worth trying to figure out how that can be avoided.

It'd be nice to have one workspace environment that handles all ESP32 boards, just as we do now for ARM.  (RISCV tends to default to different toolchains per SOC, but those SOCs come from different vendors so that's a little easier to justify.)

But if we would have to switch, it's likely to be important to make it easier to do so.  One option could be to have the user source `${HOME}/.expressif/zephyr-esp32.sh` (`...-esp32s2.sh`, ...) to set `EXPRESSIF_TOOLCHAIN_PATH` correctly, where those setup scripts could be generated by `west expressif update`.

That can be deferred, though by doing it now you'd eliminate the need to get the installation-specific toolchain version element `esp-2020r3-8.4.0` into the environment variable.

pabigot replied:
> 1. Write down on the getting started tutorial what `west espressif update` will perform in details;

For that you could just point to the upstream documentation, and say "`west espressif update` will execute steps 1, 2, 3, 4 (?) of (the upstream documentation)", with a *short* summary of what that means.  No need to go into details that are documented elsewhere.

sylvioalves replied:
Regarding the toolchain path, we are working to improve that in a way no switching will be required among SoCs. However, this might take a few weeks. Can we keep current approach for now?

pabigot replied:
Absolutely, yes.

sylvioalves replied:
@pabigot, 
I have updated the getting started document and installation process. Let me know how that sounds.

- User can either download ESP32 toolchain manually or run `west` command to automatically retrieve it.

- Added the download path, export statement and notes so that user won't be in the dark with the installation environment.

Regarding other ESP32 SoC support, we will handle that in a another moment. For now, let's work with this so that pending ESP32 features can be merged as well.

sylvioalves commented:
@ExtremeGTX @dcpleung, can you please also review?

ExtremeGTX commented:
> @ExtremeGTX @dcpleung, can you please also review?

Unfortunately I'm on vacation now, I will come back on 11 Jan. and will review the changes if it is not yet merged.

sylvioalves commented:
@nashif @carlescufi @mbolivar-nordic I have seen the slack channel talking about west espressif/sof update regarding submodules and binaries. WIll that have any impact on this PR? How can we move on on this?

pabigot commented on soc/xtensa/esp32/soc.h:
```diff
@@ -24,9 +24,7 @@ static inline void esp32_clear_mask32(uint32_t v, uint32_t mem_addr)
-extern int esp32_rom_intr_matrix_set(int cpu_no,
-				     int interrupt_src,
-				     int interrupt_line);
+extern void intr_matrix_set(int cpu_no, uint32_t model_num, uint32_t intr_num);
```
Please retain the `esp32_rom_` prefix here to avoid contaminating the global namespace with symbols other modules might define.  See another comment about having two exposed symbols; Zephyr code should use the one with the prefix.

pabigot commented on boards/xtensa/esp32/esp32.dts:
```diff
@@ -24,11 +24,11 @@
 
 &cpu0 {
-	clock-frequency = <ESP32_CLK_CPU_240M>;
+	clock-frequency = <ESP32_CLK_CPU_80M>;
```
The commit message subject and body seem to be out of sync, and there's a lot of other changes in here that don't seem to be directly related to XIP support, building a bootloader, or disabling a WDT.  Like this one.

Could you break this down into more targeted commits?

Also the text about disabling the bootloader-enabled WDT seems to conflict with what's in the third commit, where the WDT is still enabled if that wasn't done by the bootloader (where some alias variables are moved).

sylvioalves replied:
> Also the text about disabling the bootloader-enabled WDT seems to conflict with what's in the third commit, where the WDT is still enabled if that wasn't done by the bootloader (where some alias variables are moved).

Not so clear for me. Current code performs this:

1. If board starts **without**esp-idf bootloader, it disables watchdog timer used on 1st bootloader (rom).
2. If board starts **with** bootloader, it means 2nd bootloader will trigger RTC WDT to perform some checks. Then we disable that before zephyr starts.

sylvioalves replied:
@pabigot I've pushed new changes based on all discussion. Let me know if this one is still valid.

pabigot commented on boards/xtensa/esp32/esp32.dts:
```diff
@@ -24,11 +24,11 @@
 
 &cpu0 {
-	clock-frequency = <ESP32_CLK_CPU_80M>;
+	clock-frequency = <ESP32_CLK_CPU_240M>;
```
But wasn't this "bug" introduced in the first commit?  Fix it there, please.

sylvioalves replied:
What do you mean by fix it here?

Once bootloader and XIP was taken in place on 1st commit, system clock was not working for frequencies > 80MHz. That is the reason that on 1st commit it was changed to 80MHz as default.
This commit fixes it and reset default clock rate to 240MHz.

pabigot replied:
> Yes, you have a point. I'll fix in there, keep 240MHz as default.

Well, I won't insist on that, if doing that breaks builds with the intervening commits (e.g. when doing git bisect).  An acceptable alternative is to document in the first commit why it's being changed to 80 MHz, noting that that will be fixed after other changes occur.

But if you can avoid the breakage by refactoring a little that'd be fine too.

sylvioalves replied:
I squashed both commits into the 1st one. It means 1st commit regarding bootloader and XIP stuff will have also clock source changed. Sounds good?

pabigot replied:
I'll have to see it (and how the commit comment is updated), but if the devicetree content doesn't change I wouldn't notice there was a problem, so it's probably fine.

sylvioalves replied:
There are a few changes on 1st commit regarding code style.. Can we keep it?

pabigot replied:
It'd certainly be preferable to not mix whitespace and line wrapping changes in with things that have an actual effect on behavior.  I'm not sure I'd insist on it, but others would, so it could be worth dropping the changes that are cosmetic-only.

sylvioalves replied:
Removed all comestic stuff.

pabigot commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -38,13 +40,33 @@ PROVIDE ( esp32_rom_i2c_writeReg = 0x400041a4 );
+PROVIDE ( roundup2 = 0x4000ab7c );
+PROVIDE ( ets_delay_us = 0x40008534 );
+PROVIDE ( crc32_le = 0x4005cfec );
+PROVIDE ( g_ticks_per_us_pro = 0x3ffe01e0 );
```
These also should be prefixed, with an alias for `esp-idf` backwards compatibility if necessary.  Zephyr should only access these with the `esp32` prefix.

pabigot reviewed (changes requested):
I think this'll be a big improvement.  Some comments, the most significant one being that all references in Zephyr to material provided in the ROM should use symbols that are prefixed with something to deconflict them from other Zephyr modules.  If `esp-idf` still references the ROM content with a non-prefixed identifier that's fine until it gets cleaned up, but in that case provide two symbols and don't use the non-prefixed identifier in Zephyr code.

sylvioalves commented:
@pabigot this new push includes:
- added `esp32_rom` prefix in all zephyr's calls. Future esp releases already have this `esp_rom` prefix and we shall update linker file soon.
- Squashed 3rd commit (fix clock) into 1st commit so that DTS changes won't take place as it introduced a "bug".

pabigot commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -29,6 +31,10 @@ PROVIDE ( intr_matrix_set = 0x4000681c );
 PROVIDE ( g_ticks_per_us_app = 0x3ffe40f0 );
 PROVIDE ( g_ticks_per_us_pro = 0x3ffe01e0 );
 PROVIDE ( ets_delay_us = 0x40008534 );
+PROVIDE ( gpio_output_set = 0x40009b24 );
```
These are part of a block that I think should be documented as to why these have to be global symbols.  Are they all things that will become unnecessary when `esp-idf` is cleaned up?

sylvioalves replied:
These symbols are used on `hal_espressif` external module branch (i.e. esp-idf). 
To get rid of this "no prefix" symbols, it would require to change the whole esp-idf source as well. In fact, we will change it soon and the majority of the symbols will have `esp_rom` prefix (not `esp32_rom`) in it. Then, this `linker.ld` file won't have these "no prefix" anymore.
I can add a line comment on the beginning of these block, sounds good?
By the way, Is there any guideline regarding linker PROVIDE naming rule?

pabigot replied:
> In fact, we will change it soon and the majority of the symbols will have `esp_rom` prefix (not `esp32_rom`) in it.

That sounds fine.  When that happens update all the Zephyr source with s/esp32_rom/esp_rom/ so everything's consistent.

> I can add a line comment on the beginning of these block, sounds good?

Yes.  Thank you.

> By the way, Is there any guideline regarding linker PROVIDE naming rule?

Probably not.  It's implicit in rules that global symbols should be sufficiently unique to prevent conflict with other Zephyr modules and external libraries that take similar care.  Those rules are probably also unwritten, but should fall under #29569.

pabigot commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -186,67 +221,52 @@ SECTIONS
 
-    __log_backends_start = .;
+    . = ALIGN(4);
+    __esp_log_backends_start = .;
```
I don't think this change is correct.  The logging backend is looking for the unprefixed symbols.

pabigot commented on soc/xtensa/esp32/linker.ld:
```diff
@@ -276,8 +296,88 @@ SECTIONS
+/* Restore original value for symbols referenced by `common-rom.ld` */
+__log_const_start = __esp_log_const_start;
+__log_const_end = __esp_log_const_end;
+__log_backends_start = __esp_log_backends_start;
```
Oh, they're coming back here.  This seems a little complex.  I guess it's necessary.

pabigot reviewed (commented):
A few more comments.  I haven't tried building this yet, will do before approving.

The real changes are a lot easier to see with the cosmetic changes masked (hide whitespace changes in the github review page).

pabigot commented on drivers/entropy/entropy_esp32.c:
```diff
@@ -8,52 +8,77 @@
+
+	do {
+		ccount = XTHAL_GET_CCOUNT();
+		result ^= REG_READ(WDEV_RND_REG);
```
This is going off on a tangent, but I'm not sure I trust this.

Can this loop ever execute only once?  E.g. if it's been a long time since the last time it was invoked, so the cycle difference is large?

And if it does, is it possible that `WDEV_RND_REG` might not be changed between the XOR in the loop and the XOR on exit?

If so, this function would seem to return zero more often than expected.

sylvioalves replied:
> Can this loop ever execute only once? E.g. if it's been a long time since the last time it was invoked, so the cycle difference is large?

Yes, it will execute once most of the time where current call timestamp is 16 APB clocks longer than last call. Otherwise, it loops a few times.

> And if it does, is it possible that WDEV_RND_REG might not be changed between the XOR in the loop and the XOR on exit?

WDEV_RND_REG shall not return same value on subsequent readings, not causing the zero result output.

pabigot replied:
If the register is completely different each time it's read, but you just need the delay to make sure the required entropy has been incorporated, this logic would seem to be more clear:
```
	static uint32_t last_ccount;
	uint32_t ccount;
	uint32_t result;

	do {
		ccount = XTHAL_GET_CCOUNT();
		result = REG_READ(WDEV_RND_REG);
	} while (ccount - last_ccount < cpu_to_apb_freq_ratio * 16);

	last_ccount = ccount;
	return result;
```
I get concerned about comments like "we'll xor the value but it's probably not necessary", as well as the repeated XORing if the value is supposed to be 32-bits of random. Extra manipulation doesn't seem to increase the randomness.

sylvioalves replied:
@pabigot, this implementation came from esp-idf and has a background history (...). After checking internally with the team, it makes sense to dismiss the extra manipulation as you said. Thanks.

pabigot commented on drivers/entropy/entropy_esp32.c:
```diff
@@ -8,52 +8,77 @@
+		uint32_t word = entropy_esp32_get_u32();
+		uint32_t to_copy = MIN(sizeof(word), len);
+
+		memcpy(buf_bytes, &word, to_copy);
```
You can use `memmove` here since `buf_bytes` and `word` can't overlap.  It might be faster (assuming the compiler doesn't detect they're disjoint).

pabigot reviewed (approved):
This looks much better; thank you.  I've tested this with hello world and it works in my environment with only one change to my setup (explicitly set the path, rather than source `esp-idf/export.sh` from my `~/esp` directory).

I have a concern about the entropy collection, but not enough to withhold approval.

pabigot commented:
@tejlmand @nashif  please revisit.  (Did we loose "re-request review" functionality in github?  I couldn't find how to do that.)

tejlmand reviewed (commented):
LGTM
